klippa/src/lib.rs

```diff
@@ -7,67 +7,216 @@ mod parsing_util;
 pub use parsing_util::{parse_unicodes, populate_gids};
 
 use int_set::IntSet;
-use std::collections::BTreeSet;
+use skrifa::MetadataProvider;
+use std::collections::{BTreeSet, HashMap};
 use std::path::PathBuf;
 use thiserror::Error;
-use write_fonts::read::{FontRef, TableProvider, TopLevelTable};
+use write_fonts::read::{
+    tables::glyf::{Glyf, Glyph},
+    tables::loca::Loca,
+    FontRef, TableProvider, TopLevelTable,
+};
 use write_fonts::types::GlyphId;
 use write_fonts::types::Tag;
 use write_fonts::{
     from_obj::FromTableRef, tables::hhea::Hhea, tables::hmtx::Hmtx, tables::maxp::Maxp, FontBuilder,
 };
+
+const MAX_COMPOSITE_OPERATIONS_PER_GLYPH: u8 = 64;
+const MAX_NESTING_LEVEL: u8 = 64;
+const CODEPOINT_INVALID: u32 = 0xFFFFFFFF;
```

Comment:
I find this constant a bit confusing; there are many invalid codepoints before this one, and it looks like it is only used to construct a `GlyphId`? It might be simpler to just do,

```rust
const MAX_GID: GlyphId = GlyphId::new(u16::MAX);
```